iree-compile crash / codgen error involving vector.gather op

### What happened?

Following on from the resolved #11062 (thank you for fixing!), there is another compiler crash:

```
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/numerics.py:37:0: error: 'vector.gather' op operand #2 must be vector of integer or index values, but got 'index'
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/operators.py:146:0: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
/Users/___/Library/Python/3.9/lib/python/site-packages/veros/core/operators.py:146:0: error: failed to serialize executables
compilation failed
```

```
Invoked with:
 iree-compile /Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false
```


### Steps to reproduce your issue

```core-command-line.txt```:
```
/Users/___/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/wiggles/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false --mlir-pass-pipeline-crash-reproducer=savedir/core-reproducer.mlir
```

reproducer from ```core-reproducer.mlir```:

[core-reproducer.mlir.txt](https://github.com/iree-org/iree/files/9988026/core-reproducer.mlir.txt)


### What component(s) does this issue relate to?

Compiler

### Version information

[iree candidate candidate-20221110.323](https://github.com/iree-org/iree/releases/tag/candidate-20221110.323)

### Additional context

_No response_

--- a/compiler/src/iree/compiler/InputConversion/MHLO/Passes.cpp
+++ b/compiler/src/iree/compiler/InputConversion/MHLO/Passes.cpp
@@ -53,6 +53,7 @@ void registerMHLOConversionPassPipeline() {
 // Prepare HLO for use as an input to the Flow dialect.
 void buildMHLOInputConversionPassPipeline(OpPassManager &passManager) {
   passManager.addPass(mlir::mhlo::createStablehloLegalizeToHloPass());
+  passManager.addNestedPass<func::FuncOp>(mlir::createCanonicalizerPass());
   passManager.addNestedPass<func::FuncOp>(
       mhlo::createLegalizeControlFlowPass());
 

